pkg/providers/vsphere/vmlifecycle/bootstrap_cloudinit.go

```diff
@@ -49,17 +49,55 @@ func BootStrapCloudInit(
 		sshPublicKeys = strings.Join(cloudInitSpec.SSHAuthorizedKeys, "\n")
 	}
 
-	uid := cloudInitSpec.InstanceID
-	if uid == "" {
-		// InstanceID will be set to the VM bios uuid when creating new VMs.
-		// Use VM.UID for backward compatibility when InstanceID is not set.
-		uid = string(vmCtx.VM.UID)
-	}
-	if value, ok := vmCtx.VM.Annotations[vmopv1.InstanceIDAnnotation]; ok {
-		uid = value
-	}
+	// The Cloud-Init instance ID is from spec.bootstrap.cloudInit.instanceID.
+	iid := cloudInitSpec.InstanceID
+
+	if iid == "" {
+		// For new VM resources that use the Cloud-Init bootstrap provider, if
+		// the field spec.bootstrap.cloudInit.instanceID is empty when the VM
+		// resource is created, the field defaults to the value of
+		// spec.biosUUID.
+		//
+		// However, to maintain backwards compatibility with older VM resources
+		// that existed prior to the introduction of the field
+		// spec.bootstrap.cloudInit.instanceID, use the VM resource's object
+		// UID as the Cloud-Init instance ID if the value of
+		// spec.bootstrap.cloudInit.instanceID is empty.
+		iid = string(vmCtx.VM.UID)
+	}
+
+	if v := vmCtx.VM.Annotations[vmopv1.InstanceIDAnnotation]; v != "" && v != iid {
+		// Since older VMs used the VM resource's object UID for the Cloud-Init
+		// instance ID, it was necessary to preserve the original Cloud-Init
+		// instance ID if the VM was backed up and restored at a later date.
+		// This is because restored VM resources are not the same Kubernetes
+		// resource, and therefore have new object UIDs.
+		//
+		// A new object UID would have meant a new Cloud-Init instance ID, and
+		// the latter means that Cloud-Init would have run all of the
+		// per-instance modules again, this behavior needed to be mitigated.
+		// Therefore, the restore workflow introduced the
+		// vmopv1.InstanceIDAnnotation in order to override the value of the
+		// Cloud-Init instance ID in case it needed to be something *other* than
+		// the Kubernetes object UID.
+		//
+		// If the instance ID from the annotation is non-empty *and* is
+		// different than the current instance ID value (whether it is from
+		// spec.bootstrap.cloudInit.instanceID or metadata.uid), go ahead and
+		// use the value from the annotation.
+		iid = v
+		delete(vmCtx.VM.Annotations, vmopv1.InstanceIDAnnotation)
+	}
```

Comment:
The restore process actually does the opposite of what these comments describe. That is, if a backed-up VM has this annotation set, the restore service **deletes** this annotation before creating the VM in a K8s cluster. As you mentioned, a new instance ID (from the new object UID) ensures Cloud-Init runs all of the per-instance modules, including the network, so that the restored VM will get the correct new IP upon restore.

The actual use case of this annotation is in the upgrade process, which sets to `iid-datasource-none`. It addresses a previous bug when a VM is deployed with Cloud-Init without user data. See this PR that initially introduced this annotation: https://github.com/vmware-tanzu/vm-operator/pull/211 (I am sure you have the context since you proposed this fix. I am just posting it here for others' reference).

Since we have `cloudInitSpec.InstanceID` now, I think we could retire this annotation after updating our internal upgrade script to use this spec to set the `iid-datasource-none` to fix the original bug.

Reply:
@dilyar85 So really it's my comments that are wrong, but the logic is fine? Let me update the comments. Thank you!

Reply:
@dilyar85 Can you please review the new comments for accuracy? Thanks!